Mathlib/Analysis/Normed/Module/Complemented.lean

```diff
@@ -128,4 +128,119 @@ theorem ClosedComplemented.of_quotient_finiteDimensional [CompleteSpace 𝕜]
+      ContinuousLinearEquiv.coe_coe, coe_prodEquivOfClosedCompl,
+      coe_continuous_linearProjOfClosedCompl', Function.comp_apply,
+      (linearProjOfIsCompl_apply_left h ⟨x,hx⟩), ContinuousLinearMap.inl_apply,
+      coe_prodEquivOfIsCompl', ZeroMemClass.coe_zero, add_zero]
+  · intro hx
+    simp [idempotentOfClosedCompl] at hx
+    rw [← hx]
+    exact coe_mem ((linearProjOfIsCompl p q h) x)
+
+lemma mem_iff_zero_ofClosedCompl (h : IsCompl p q) (hp : IsClosed (p : Set E))
+    (hq : IsClosed (q : Set E)) {y : E} :
+    let P := idempotentOfClosedCompl p q h hp hq
+    y ∈ q ↔ P y = 0 := by
+  constructor
+  · intro hy
+    simp only [idempotentOfClosedCompl, ContinuousLinearMap.coe_comp',
+      ContinuousLinearEquiv.coe_coe, coe_prodEquivOfClosedCompl,
+      coe_continuous_linearProjOfClosedCompl', Function.comp_apply, ContinuousLinearMap.inl_apply,
+      coe_prodEquivOfIsCompl', ZeroMemClass.coe_zero, add_zero, ZeroMemClass.coe_eq_zero,
+      ((linearProjOfIsCompl_apply_eq_zero_iff h).mpr hy)]
+  · intro h
+    rw [idempotentOfClosedCompl] at h
+    simp at h
+    exact h
+
+lemma is_idempotent_ofClosedCompl (h : IsCompl p q) (hp : IsClosed (p : Set E))
+    (hq : IsClosed (q : Set E)) :
+    let P := idempotentOfClosedCompl p q h hp hq
+    IsIdempotentElem P := by
+  ext z
+  have hy1 : z ∈ p ⊔ q := by
+    rw [h.sup_eq_top]
+    exact AddSubgroup.mem_top z
+  obtain ⟨x₁,⟨hx₁,⟨y₁,⟨hy₁,hx₁y₁y⟩⟩⟩⟩ := Submodule.mem_sup.mp hy1
+  rw [← hx₁y₁y, map_add, map_add, ((mem_iff_zero_ofClosedCompl h hp hq).mp hy₁),
+    ((mem_iff_invariant_ofClosedCompl h hp hq).mp hx₁), add_zero,
+    ContinuousLinearMap.coe_mul, Function.comp_apply, Function.comp_apply,
+    ((mem_iff_zero_ofClosedCompl h hp hq).mp hy₁),
+    ((mem_iff_invariant_ofClosedCompl h hp hq).mp hx₁),
+    map_zero, add_zero, (mem_iff_invariant_ofClosedCompl h hp hq).mp hx₁]
+
+lemma ker_idempotentOfClosedCompl (h : IsCompl p q) (hp : IsClosed (p : Set E))
+    (hq : IsClosed (q : Set E)) :
+    let P := idempotentOfClosedCompl p q h hp hq
+    ker P = q := by
+  ext x
+  simp only [idempotentOfClosedCompl, LinearMap.mem_ker, ContinuousLinearMap.coe_comp',
+    ContinuousLinearEquiv.coe_coe, coe_prodEquivOfClosedCompl,
+    coe_continuous_linearProjOfClosedCompl', Function.comp_apply, ContinuousLinearMap.inl_apply,
+    coe_prodEquivOfIsCompl', ZeroMemClass.coe_zero, add_zero, ZeroMemClass.coe_eq_zero,
+    linearProjOfIsCompl_apply_eq_zero_iff]
+
+lemma range_idempotentOfClosedCompl (h : IsCompl p q) (hp : IsClosed (p : Set E))
+    (hq : IsClosed (q : Set E)) :
+    let P := idempotentOfClosedCompl p q h hp hq
+    range P = p := by
+  ext x
+  exact ⟨fun ⟨y, hy⟩ => by simp [idempotentOfClosedCompl, ← hy],
+    fun hx => LinearMap.mem_range.mp ⟨x,(mem_iff_invariant_ofClosedCompl h hp hq).mp hx⟩⟩
+
+omit [CompleteSpace E] in
+lemma ker_id_sub_eq_range {P : E →L[𝕜] E} (h : IsIdempotentElem P) : ker (1 - P) = range P := by
```

Comment:
The `omit` is a clue that this result (and the next) don't belong here. You should move them and generalize.

Reply:
Oops, forgot to come back to fix that, sorry.